### PR TITLE
more general check_matching_dims

### PR DESCRIPTION
**Pull request by nhuurre**
## Summary

Generalize `check_matching_dims` so it can handle arbitrary containers and not just matrices. All dimensions must match for deep containers.

## Tests

Expanded `check_matching_dims` unit tests to cover more cases.

## Side Effects

None

## Release notes

Extended check_matching_dims to handle arbitrary containers.

## Checklist

- [X] Math issue #1935

- [X] Copyright holder: Niko Huurre

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

File: stan/math/prim/err/check_matching_dims.hpp [hunk 1]
```diff
@@ -70,6 +70,43 @@ inline void check_matching_dims(const char* function, const char* name1,
...
+ * @param name2 variable name for the second container (for error messages)
+ * @param y2 second container to test
+ * @throw <code>std::invalid_argument</code> if the dimensions of the
+ *    containers do not match
+ */
+template <typename T1, typename T2>
+inline void check_matching_dims(const char* function, const char* name1,
+                                const std::vector<T1>& y1, const char* name2,
+                                const std::vector<T2>& y2) {
+  check_size_match(function, name1, y1.size(), name2, y2.size());
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_matching_dims.hpp:105]: You could simplify this by using `dims()`: https://github.com/stan-dev/math/blob/develop/stan/math/prim/fun/dims.hpp

No need for the scalar overload then, just call dims and iterate over the resulting vector.

I know there is not much code here, so not a huge deal, but still makes sense not to duplicate.

Comment (nhuurre) [reply stan/math/prim/err/check_matching_dims.hpp:105]: You're right. I wonder why I didn't do that... I think I wanted to preserve `rows of`/`columns of` in the error message when the base type is a matrix. But the error messages have are half-baked anyway...

Comment (rok-cesnovar) [reply stan/math/prim/err/check_matching_dims.hpp:105]: I think with calling dims the error message might actually be better as you could mention on which dimension sizes differ.

**Comment by rok-cesnovar** [comment]
There is an issue with overloaded dims functions that clashes with the overload in /opencl. Yack. I will push a fix here if you dont mind and then let @t4c1 review since this PR will include my changes also.

**Pull request merged**